onnxruntime/core/graph/contrib_ops/nhwc_inference_context.h

```diff
@@ -0,0 +1,128 @@
+    return ctx_.getAttribute(name);
+  }
+
+  size_t getNumInputs() const noexcept override {
+    return ctx_.getNumInputs();
+  }
+
+  const ONNX_NAMESPACE::TypeProto* getInputType(size_t index) const override {
+    return (index == 0) ? &input_type_ : ctx_.getInputType(index);
+  }
+
+  const ONNX_NAMESPACE::TensorProto* getInputData(size_t index) const override {
+    // we can't return the NHWC input data without transposing it, but wouldn't expect to be asked for it
+    // during shape inferencing as getInputData is only used to retrieve things that may have small 
+	// constant initializers (e.g. something like the min and max values of a Clip operator).
+    return index == 0 ? nullptr : ctx_.getInputData(index);
+  }
+
+  size_t getNumOutputs() const noexcept override {
+    return ctx_.getNumOutputs();
+  }
+
+  ONNX_NAMESPACE::TypeProto* getOutputType(size_t index) override {
+    return (index == 0) ? &output_type_ : ctx_.getOutputType(index);
+  }
+
+  ONNX_NAMESPACE::GraphInferencer* getGraphAttributeInferencer(const std::string& attribute_name) override {
+    return ctx_.getGraphAttributeInferencer(attribute_name);
+  }
+
+  const ONNX_NAMESPACE::TensorShapeProto* getSymbolicInput(size_t index) const override {
+    return ctx_.getSymbolicInput(index);
+  }
+
+  const ONNX_NAMESPACE::SparseTensorProto* getInputSparseData(size_t index) const override {
+    return ctx_.getInputSparseData(index);
+  }
+
+  // Propagate the inferred type/shape info to output 0, converting any inferred shape from NCHW to NHWC.
+  void PropagateOutputShape() {
+    auto& nhwc_tp = *ctx_.getOutputType(0);
+
+    // copy latest type/shape info.
+    nhwc_tp = output_type_;
+
+    // convert shape to channels last
+    if (output_type_.tensor_type().has_shape()) {
+      const auto& nchw_shape = output_type_.tensor_type().shape();
+      const int rank = nchw_shape.dim_size();
+      // N and C dims are required. Some operators like AveragePool allow 1D input
+      if (rank < 3) {
+        fail_shape_inference("Output tensor must have at least 3 dimensions");
+      }
+
+      // Convert output shape from N, C, H {, W, ...} to N, H {, W, ...}, C.
+      auto& nhwc_shape = *nhwc_tp.mutable_tensor_type()->mutable_shape();
+      nhwc_shape.Clear();
+      *nhwc_shape.add_dim() = nchw_shape.dim(0);
+      for (int i = 2; i < rank; i++) {
+        *nhwc_shape.add_dim() = nchw_shape.dim(i);
+      }
+
+      *nhwc_shape.add_dim() = nchw_shape.dim(1);
+    }
+  }
+
+ private:
+  void TransposeToNchw(const ONNX_NAMESPACE::TypeProto& nhwc_tp, ONNX_NAMESPACE::TypeProto& nchw_tp) {
+    if (nhwc_tp.tensor_type().has_shape()) {
+      const auto& nhwc_shape = nhwc_tp.tensor_type().shape();
+      const int rank = nhwc_shape.dim_size();
+      // N and C dims are required. Some operators like AveragePool allow 1D input
+      if (rank < 3) {
+        fail_shape_inference("Input tensor must have at least 3 dimensions");
```

Comment:
>Input [](http://example.com/codeflow?start=30&length=5)

nit: TransposeToNchw() is also called for output 0, could "Input tensor" be confusing?